contracts/data-storage/andromeda-graph/src/contract.rs

```diff
@@ -0,0 +1,305 @@
+        &ctx.info,
+        &ctx.env,
+        &ctx.amp_ctx,
+        msg.as_ref(),
+    )?;
+
+    let res = match msg {
+        ExecuteMsg::UpdateMap { map_info } => execute_update_map(ctx, map_info, action),
+        ExecuteMsg::StoreCoordinate {
+            coordinate,
+            is_timestamp_allowed,
+        } => execute_store_coordinate(ctx, coordinate, is_timestamp_allowed, action),
+        _ => ADOContract::default().execute(ctx, msg),
+    }?;
+
+    Ok(res
+        .add_submessages(action_response.messages)
+        .add_attributes(action_response.attributes)
+        .add_events(action_response.events))
+}
+
+pub fn execute_update_map(
+    ctx: ExecuteContext,
+    map_info: MapInfo,
+    action: String,
+) -> Result<Response, ContractError> {
+    let sender = ctx.info.sender;
+    ensure!(
+        ADOContract::default().is_owner_or_operator(ctx.deps.storage, sender.as_ref())?,
+        ContractError::Unauthorized {}
+    );
+
+    let map = MAP_INFO
+        .load(ctx.deps.storage)
+        .map_err(|_| ContractError::InvalidParameter {
+            error: Some("Map not found".to_string()),
+        })?;
+
+    ensure!(
+        map != map_info,
+        ContractError::InvalidParameter {
+            error: Some("Map already exists".to_string())
+        }
+    );
+
+    MAP_INFO.save(ctx.deps.storage, &map_info)?;
+
+    let max_point = POINT.load(ctx.deps.storage)?;
+
+    for point in 1..=max_point {
+        MAP_POINT_INFO.remove(ctx.deps.storage, &point);
+    }
+
+    POINT.save(ctx.deps.storage, &0)?;
+
+    Ok(Response::new().add_attributes(vec![attr("action", action), attr("sender", sender)]))
+}
+
+pub fn execute_store_coordinate(
+    ctx: ExecuteContext,
+    coordinate: Coordinate,
+    is_timestamp_allowed: bool,
+    action: String,
+) -> Result<Response, ContractError> {
+    let sender = ctx.info.sender;
+    ensure!(
+        ADOContract::default().is_owner_or_operator(ctx.deps.storage, sender.as_ref())?,
+        ContractError::Unauthorized {}
+    );
+
+    let map = MAP_INFO
+        .load(ctx.deps.storage)
+        .map_err(|_| ContractError::InvalidParameter {
+            error: Some("Map not found".to_string()),
+        })?;
+
+    let MapInfo {
+        map_size,
+        allow_negative,
+        map_decimal,
+    } = map;
+    let x_length = map_size.x_width as f64;
+    let y_length = map_size.y_width as f64;
+    let z_length = map_size.z_width.map(|z| z as f64);
+
+    let is_z_allowed = z_length.is_some();
+
+    let x_coordinate = ((coordinate.x_coordinate * 10_f64.powf(map_decimal as f64)) as i64) as f64
+        / 10_f64.powf(map_decimal as f64);
+    let y_coordinate = ((coordinate.y_coordinate * 10_f64.powf(map_decimal as f64)) as i64) as f64
+        / 10_f64.powf(map_decimal as f64);
+    let z_coordinate = coordinate.z_coordinate.map(|z| {
+        ((z * 10_f64.powf(map_decimal as f64)) as i64) as f64 / 10_f64.powf(map_decimal as f64)
+    });
+
+    match z_coordinate {
```

Comment:
Can't this be done in a single boolean check?

```z_coordinate.is_some() == is_z_allowed```

Reply:
It was resolved.